gateway/app/drivers/qodProvisioning/provisioning.py

```diff
@@ -0,0 +1,284 @@
+            sink=sub_info.sink,
+            sinkCredential=sub_info.sinkCredential,
+            device=sub_info.device,
+            provisioningId=sub_info.provisioningId,
+            startedAt=sub_info.startedAt,
+            status=status,
+            statusInfo=StatusInfo.NETWORK_TERMINATED
+            if status == Status.UNAVAILABLE
+            else None,
+        )
+
+        return provisioning_info
+
+    async def send_callback_message(
+        self, body: UserPlaneNotificationData, id: str
+    ) -> None:
+        key = f"{_prefix_info}:{id}"
+        sub_info = ProvisioningInfo.model_validate_json(await self.redis.get(key))
+
+        if sub_info is None:
+            raise ResourceNotFound()
+
+        sub_info.status = self._get_qod_status(body)
+
+        # TODO: change this to cloud event
+        cloud_event = CloudEvent(
+            id=str(uuid.uuid4()),
+            source=f"{settings.provisioning.gateway_url}/device-qos/{id}",
+            type=Type.org_camaraproject_qod_provisioning_v0_status_changed,
+            specversion=Specversion.field_1_0,
+            datacontenttype=Datacontenttype.application_json,
+            data=sub_info.model_dump(mode="json"),
+            time=datetime.datetime.now(),
+        )
+
+        sink = sub_info.sink
+        if sink is None:
+            raise Exception()
+
+        await self.httpx_client.post(sink, content=cloud_event.model_dump_json())
+
+    def _get_subscription_id_from_subscription_url(self, url: AnyUrl | None) -> str:
+        if url is None or url.path is None:
+            raise Exception("Invalid url")
+
+        return url.path.rsplit("/")[-1]
+
+    def _get_phone_number(self, gpsi: str | None) -> str | None:
+        if gpsi is None:
+            return None
+
+        phone_number = gpsi.strip("msisdn-")
+        return phone_number
+
+    def _get_qod_status(
+        self, nef_response: UserPlaneNotificationData | AsSessionWithQoSSubscription
+    ) -> Status:
+        if isinstance(nef_response, UserPlaneNotificationData):
+            if nef_response.eventReports[0].event == UserPlaneEvent.QOS_GUARANTEED:
+                return Status.AVAILABLE
+            elif (
+                nef_response.eventReports[0].event == UserPlaneEvent.SESSION_TERMINATION
+            ):
+                return Status.UNAVAILABLE
+            elif nef_response.eventReports[0].event == UserPlaneEvent.QOS_MONITORING:
+                return Status.REQUESTED
+        elif (
+            isinstance(nef_response, AsSessionWithQoSSubscription)
+            and nef_response.events
+        ):
+            if nef_response.events[0] == UserPlaneEvent.QOS_GUARANTEED:
+                return Status.AVAILABLE
+            elif nef_response.events[0] == UserPlaneEvent.SESSION_TERMINATION:
+                return Status.UNAVAILABLE
+            elif nef_response.events[0] == UserPlaneEvent.QOS_MONITORING:
+                return Status.REQUESTED
```

Comment:
Unfortunately not how it works. The `events` property of the `AsSessionWithQoSSubscription` is used to specify the events you want to subscribe to, it doesn't track issued events.

You need to track in the database the status from all received notifications